sysbuild/west: runners: Add optional configuration for setting flashing configuration

nordicjm opened the pull request:
This set of commits adds optional configuration option to boards when flashing sysbuild-enabled images to allow for:

- Setting board priorities, which boards should be flashed before others
- Preventing reset after flashing each image and resetting only once all images have been flashed
- Preventing commands specified on the command line to `west flash` running for each image flash stage

These features are opt-in and disabled by default, configurations are added for nrf5340dk_nrf5340, thingy53_nrf5340 and nrf9160dk_nrf9160 boards which addresses some possible issues with flashing multiple images on these boards. The main issues that this resolves are:

- Security being enabled by an image, if one of the images enables security on the micro e.g. mcuboot, then without this, later images would fail to flash
- Running "west flash --recover" recovering the flash each time an image is flash, wrongly erasing the flash contents from the previous image flash invocations
- nRF5340 recovery for the network core recovering both application and network cores, whilst recovery for the application core only impacting the application core

Also fixes #53487 by fixing the wrong argument name used by the stm32cubeprogrammer west flash test script

nordicjm commented:
Actually thinking about this, this will not work for _ns targets

joerchan commented:
Can you elaborate?

nordicjm committed:
west: runners: Add single-use commands per board flash

This adds supports for flashing images with sysbuild where there
are multiple images per board to prevent using the same command per
image flash which might cause issues if they are not ran just once
per flash per unique board name. This can be enabled per-board by
creating a `flash_runner.yml` file in the board directory with a
list named `board_single_commands`, e.g. to prevent the --recover
command being used multiple times when flashing a board, the file
contents would be similar to:

board_single_commands:
  - --recover: [
      nrf9160dk_nrf9160,
      nrf9160dk_nrf9160_ns,
    ]

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm committed:
boards: arm: nrf5340dk_nrf5340: Prevent multiple recover use in flash

Prevents the --recover argument for west flash being used multiple
times when flashing a sysbuild image for the nRF5340 development kit,
which would otherwise cause unbootable images.

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm committed:
boards: arm: thingy53_nrf5340: Prevent multiple recover use in flash

Prevents the --recover argument for west flash being used multiple
times when flashing a sysbuild image for the thingy53, which would
otherwise cause unbootable images.

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm committed:
boards: arm: nrf9160dk_nrf9160: Prevent multiple recover use in flash

Prevents the --recover argument for west flash being used multiple
times when flashing a sysbuild image for the nRF9160 development kit,
which would otherwise cause unbootable images.

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm committed:
west: runners: Add flashing priority per board

This adds supports for flashing images with sysbuild where there are
multiple images per board in a given priority to to prevent issues
where flashing one board/image prior to another one would cause
issues with the flashing process.

This can be enabled per-board by creating a `flash_runner.yml` file
in the board directory with a list named `board_flash_priority`,
e.g. to flash an nRF5340 board in order of network core, application
core (secure) and application core (non-secure), the file contents
would be similar to:

board_flash_priority:
  - nrf5340dk_nrf5340_cpunet
  - nrf5340dk_nrf5340_cpuapp
  - nrf5340dk_nrf5340_cpuapp_ns

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm committed:
boards: arm: nrf5340dk_nrf5340: Flash network core images first

Sets up a board flashing priority order to flash the nRF5340dk's
network core first, followed by the application core (secure) and
finally application core (non-secure), this is because if a
--recover argument is provided when flashing to the network core
image, this will recover both cores, not just the network core.

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm committed:
boards: arm: thingy53_nrf5340: Flash network core images first

Sets up a board flashing priority order to flash the thingy53's
network core first, followed by the application core (secure) and
finally application core (non-secure), this is because if a
--recover argument is provided when flashing to the network core
image, this will recover both cores, not just the network core.

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm committed:
west: runners: Add option to allow/prevent board reset

This adds supports for specifying if boards should be reset or not
after an image has been flashed, this is most useful for sysbuild
where multiple images are flashed and having one boot before all
are flashed causes the debug interface to be disabled or faults,
preventing correct operation of the firmware. Existing reset
priority of runners that supported it has been maintained.

This can be enabled per-board by creating a `flash_runner.yml` file
in the board directory with a list named `board_delay_reset`, e.g.
to delay an nRF5340 board from resetting whilst network core images
are being flashed, or whilst application core images (secure and
non-secure combined) are being flashed, the file contents would be
similar to:

board_delay_reset:
  - [
      nrf5340dk_nrf5340_cpuapp,
      nrf5340dk_nrf5340_cpuapp_ns,
    ]
  - [
      nrf5340dk_nrf5340_cpunet,
    ]

Also fixes #53487 by using the correct argument for the
stm32cubeprogrammer tests of "--reset-mode" instead of "--reset".

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm committed:
boards: arm: nrf5340dk_nrf5340: Delay board reset whilst flashing

Enables the delayed board reset feature when using "west flash" with
sysbuild until all images for each core have been flashed to the
board. This prevents situations occuring where an image might enable
security protection such as locking out debug mode and then other
images failing to flash because of this.

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm committed:
boards: arm: thingy53_nrf5340: Delay board reset whilst flashing

Enables the delayed board reset feature when using "west flash" with
sysbuild until all images for each core have been flashed to the
board. This prevents situations occuring where an image might enable
security protection such as locking out debug mode and then other
images failing to flash because of this.

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm committed:
boards: arm: nrf9160dk_nrf9160: Delay board reset whilst flashing

Enables the delayed board reset feature when using "west flash" with
sysbuild until all images for each core have been flashed to the
board. This prevents situations occuring where an image might enable
security protection such as locking out debug mode and then other
images failing to flash because of this.

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm commented:
@joerchan I had an issue in the previous iteration which I've since fixed, but it wouldn't have been an issue anyway since TF-M in zephyr creates a single merged hex file which gets flashed - can you give this a try and check it works OK including with any `west flash` arguments you might use (e.g. `--recover`)? Also, would there be any plan to split TF-M up with sysbuild so that the secure and non-secure parts are flashed separately or would it always just flash a single hex file?

joerchan commented:
Our strategy has been to merge all the hex files to a single merged hex file and flash it in a single command.
So if you enable BL2 for example that would also be added to the merged hex file.
I think that is the simplest approach instead of having multiple flash commands.

From this PR this seems like a lot of files needs to be updated for all out-of-tree boards. Which would be a disadvantage for our users I think. It would have been easier if it was defined on SoC level instead of board.

I am not aware of what the plan is for TF-M and sysbuild.

nordicjm committed:
doc: release: 3.3: Add note on west flash sysbuild changes

Adds a note on changes to "west flash" usage with sysbuild for:
setting board priorities, limiting commands to run once per set
of boards and preventing a set of boards from resetting until
all images have been flashed.

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

tejlmand commented on scripts/west_commands/run_common.py:
```diff
@@ -161,6 +161,17 @@ def do_run_common(command, user_args, user_runner_args, domains=None):
+    # if there are multiple images to flash. This allows for flashing of
+    # images that e.g. setup security but allow for the remaining images to
+    # be flashed before that protection is enabled.
+    board_reset = []
```
Wonder if this should be turned around, that is that when flashing multiple images then system will not be reset until final image has been flashed.

If a reset is required after other images, then the generated `runners.yaml` can contain this information in the runner specific arguments section.

I believe that would require fewer changes.

tejlmand replied:
> it would actually just be better if all flash runners could run without doing any resetting, then once done, all cores are reset with a final invocation of each flash runner

that was exactly what I was saying, remove the auto reset we have today: `that is that when flashing multiple images then system will not be reset until final image has been flashed.`

if, for some unforeseen reason, a given image requires a reset, then it would be possible to use the function `board_runner_args(<runner> "--erase")` to forcefully request a reset for that image, even if reset would not be invoked.

Pros:
- No reset after flashing of images
- No need to keep a list of boards that should not be reset (cause there is only a single reset after all images are flashed)
- Possibility for an image to forcefully request a reset (already supported today) to accommodate special needs we don't know about atm.

If images must be flashed in a specific order, then that is a dependency tree the build system must know about, not the flash runner.

tejlmand reviewed (commented):
Thanks for taking this up.

Haven't looked detailed at the implementation itself but got a couple of general questions.

tejlmand commented:
Should some information be runner / flash tool specific defaults and not board based ?

So that `nrfjprog` default will only issue `--recover` once and not each board having to specify this information.

nordicjm commented:
> Should some information be runner / flash tool specific defaults and not board based ?
> 
> So that nrfjprog default will only issue --recover once and not each board having to specify this information.

The problem with this is linkage, for the nrf5340 if you recover the application core, it clears the application core, but if you recover the network core, it clears both cores, for another MCU it might only clear a specific core, so it's not really possible to know how to do that. SoC based would make more sense than board based because an nrf5340 is an nrf5340 no matter what board it is on, so this can be migrated to soc-level for boards v2 in the future instead of being board-specific.

nordicjm commented:
And this gets more complicated with power domains, i.e. assume there are 2 CPUs on a device, a main CPU and a co-processor, for power reduction reasons the power to the co-processor is controlled with a FET by the main CPU, if there is no image on the main CPU or it has not started executing, the second CPU cannot be flashed because there will be no power

nordicjm commented on boards/arm/thingy53_nrf5340/flash_runner.yml:
```diff
@@ -0,0 +1,9 @@
+# Recovery is only needed once per core.
+board_single_commands:
+  - --recover: [
```
I will add a new function to the runners to return what custom commands they support, and can then use that to decide if it should be used or ignored.
